### ISSUE TITLE
Passing paymentSourceId isn’t working

### ISSUE THREAD

**Issue by joshuabaker** [craftcms/commerce#1283]
### Description

It doesn’t appear possible to pass `paymentSourceId` into the `payment/pay` action as examples show. We got some generic errors whilst trying to use this and it turns out the API requests being made are missing the token (in the payment form).

Looking through the controller, there’s only [one reference of `paymentSourceId`](https://github.com/craftcms/commerce/blob/24a4d2540240f2d8116ef8739132b9d00ade360f/src/controllers/PaymentsController.php#L227) that I could find and this is only for newly created payment sources.

It’s super unclear how this works. Are we supposed to pass the token in [via the body parameters](https://github.com/craftcms/commerce/blob/24a4d2540240f2d8116ef8739132b9d00ade360f/src/controllers/PaymentsController.php#L217)? Or change our flow to update the order in a step beforehand?

### Additional info

- Craft version: 3.3.20.1
- Commerce version: 2.2.17
- PHP version: 7.3.8

**Comment by joshuabaker** [comment]
Okay, I’m gathering that the [example](https://github.com/craftcms/commerce/blob/4b5001540865fb17f8c85494a7387405c3c12da0/templates/shop/checkout/payment.html) is out of date and that it’s no longer possible to pass `paymentSourceId` to `payment/pay`.

### Example

The payment source isn’t loaded and `populateFromPaymentSource` never gets called in this example.

```twig
<form method="post">
  {{ csrfInput() }}
  {{ actionInput('commerce/payments/pay') }}
  {{ hiddenInput('gatewayId', gateway.id) }}
  {{ redirectInput('/payment/complete') }}

  {# @var paymentSource \craft\commerce\models\PaymentSource #}
  {% for paymentSource in paymentSources %}
    <div>
      <input type="radio" name="paymentSourceId" value="{{ paymentSource.id }}">
      {{ paymentSource.description }}
    </div>
  {% endfor %}
  
  <input type="submit">
</form>
```

We all expected passing `paymentSourceId` to the `payment/pay` action would update the payment form with that source.

The workaround seems to be to pass the payment form values directly (e.g. `token` instead of `paymentFormId`).

@lukeholder This doesn’t seem right. Are you able to confirm is this is the correct approach?

**Comment by joshuabaker** [comment]
No, it can’t be right. The payment form requires a customer reference too. We _have_ to call `populateFromPaymentSource`.

We should be updating the order [here](https://github.com/craftcms/commerce/blob/24a4d2540240f2d8116ef8739132b9d00ade360f/src/controllers/PaymentsController.php#L238) if the `paymentSourceId` was passed in.

```php
if ($paymentSourceId = $request->getBodyParam('paymentSourceId') !== null) {
    $order->paymentSourceId = $paymentSourceId;
}

$paymentSource = $order->getPaymentSource();
```

Adding those lines making everything work.

**Comment by joshuabaker** [comment]
Or better yet, just use the same code as the `cart/update-cart` action uses.

```php
// Set Payment Source on cart
if ($paymentSourceId = $request->getParam('paymentSourceId')) {
    $this->_cart->gatewayId = null;
    $this->_cart->paymentSourceId = $paymentSourceId;
}
```

For others encountering this, we’ve created a separate controller extending the base `PaymentsController` to set the payment source.

```php
<?php

namespace modules\payments\controllers;

use Craft;
use craft\commerce\Plugin as Commerce;

/**
 * Class PaymentsController
 *
 * @package modules\payments\controllers
 */
class PaymentsController extends \craft\commerce\controllers\PaymentsController
{
    /**
     * @return \yii\web\Response|null
     * @throws \Throwable
     * @throws \craft\errors\ElementNotFoundException
     * @throws \yii\base\Exception
     * @throws \yii\web\HttpException
     */
    public function actionPay()
    {
        // Override the pay action to add the payment source to the order
        // https://github.com/craftcms/commerce/issues/1283

        $paymentSourceId = Craft::$app->getRequest()->getBodyParam('paymentSourceId');

        if ($paymentSourceId) {
            $order = Commerce::getInstance()->getCarts()->getCart(true);
            $order->gatewayId = null;
            $order->paymentSourceId = $paymentSourceId;
            Craft::$app->getElements()->saveElement($order);
        }

        return parent::actionPay();
    }
}
```

**Comment by joshuabaker** [comment]
@lukeholder Did you have any thoughts on this?

**Comment by keyurshah** [comment]
i just encountered this issue. had to solve by updating the cart (ajax call) with the paymentSourceId before submitting. 

Your solution is a clever workaround.

**Comment by joshuabaker** [comment]
@lukeholder You mention [here](https://craftcms.stackexchange.com/a/30183) that this is possible.

**Comment by lukeholder** [comment]
Fixed in Commerce 3.1 due out this week.
